Analytics Refactor

dawoodkhan82 opened the pull request:
# Description
Use `huggingface_hub` `send_telemetry` to send analytics on interfaces and blocks.
See example of how the "blocks launch" analytics are parsed for `hello_blocks` demo. Specifically the metadata fields.(https://kibana.elastic.huggingface.tech/app/discover#/doc/60cb2590-eb2b-11ec-abc6-8728463c304a/.ds-ds-hub-telemetry-v0-2022.06.13-000001?id=SEaIlIYBuZZBmUfqkVc9).


Please include: 
* relevant motivation
* a summary of the change 
* which issue is fixed. 
* any additional dependencies that are required for this change.

Closes: # (issue)

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

dawoodkhan82 commented on requirements.txt:
```diff
@@ -2,6 +2,7 @@ aiohttp
 altair>=4.2.0
 fastapi
 ffmpy
+git+https://github.com/huggingface/huggingface_hub.git
```
will replace with`huggingface_hub` when `send_telemetry` is released.

abidlabs replied:
In `flagging.py`, we check to see if `huggingface_hub` is installed. We can now remove that check

abidlabs replied:
Also WDYT about pinning to a specific commit hash instead of installing the latest on `main` so that we have a stable, reliable install. Otherwise if anything breaks on `huggingface_hub`'s `main`, it could affect `gradio` users

dawoodkhan82 replied:
We can pin to the version of the next release of `huggingface_hub`, instead of a specific commit hash right?

abidlabs replied:
Once it's out on `huggingface_hub`, then I don't think we need to pin anymore, we can trust that `huggingface_hub` won't break on release. However, maybe I'm being paranoid, but I don't think we should assume that for `main`. Especially since our analytics happen early on, so if something breaks with analytics, it might prevent usage of gradio altogether

dawoodkhan82 commented on gradio/blocks.py:
```diff
@@ -1583,6 +1601,27 @@ def reverse(text):
                 "mode": self.mode,
             }
             utils.launch_analytics(data)
+            additional_data = {
```
adding some additional analytics here. Such as a list of event triggers and blocks components. Although what components are inputs vs outputs is lost here. Open to suggestions for more analytics to track

dawoodkhan82 replied:
Yeah combining both makes sense. I can separate blocks components into list of inputs and outputs, should be easy enough.

So we'll have `gradio/initiated`, `gradio/launched`, `gradio/error`, and `gradio/integrations` as topics

abidlabs replied:
Lists of components would make it hard to query imo. It would be nice if we can easily query things like "how often is the ColorPicker component used"? What if we have a `gradio/components` topic that stores:

* component type
* mode: input, output, both, or neither  

that both Interface and Blocks use?

freddyaboulton replied:
Would that require a separate request per component in the demo?

Im reading this as storing each component in a separate row

abidlabs replied:
Yup it would require a separate request per component as far as I understand, but if we can fire these off asynchronously in the background in `launch()`, there should be no performance overhead

abidlabs replied:
Thinking about this some more, that would be way too many network requests. Ignore my suggestion, let's stick with the topics you've proposed @dawoodkhan82! Though we should handle the fact that Blocks can have components that are both inputs and outputs, or neither.

abidlabs commented on gradio/blocks.py:
```diff
@@ -487,7 +491,8 @@ def __init__(
             else os.getenv("GRADIO_ANALYTICS_ENABLED", "True") == "True"
         )
-
+        if self.analytics_enabled == False:
```
```suggestion
        if not self.analytics_enabled:
```

abidlabs commented on gradio/blocks.py:
```diff
@@ -525,12 +530,18 @@ def __init__(
             utils.initiated_analytics(data)
 
+            # huggingface_hub send_telemetry
+            send_telemetry(
```
Is this method robust? Like if there's no network, will it raise an error?

abidlabs replied:
Also, is this a blocking function? If so, we should make this run in a background thread, like we currently do in `utils.initiated_analytics()` for example

dawoodkhan82 commented:
I addressed all the comments. For blocks launched telemetry, I now track:

(all tracked as a list of strings)
- blocks components used
- event triggers
- inputs
- outputs
- targets (for components attached to event triggers)

abidlabs commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 
+## Bug fixes
+
+- Fix bug that cause UI to be vertically centered at all times by [@pngwn](https://github.com/pngwn) in [PR 3336](https://github.com/gradio-app/gradio/pull/3336)
+- Use `huggingface_hub` to send telemetry on `interface` and `blocks`; eventaully to replace segment by [@dawoodkhan82](https://github.com/dawoodkhan82) in [PR 3342](https://github.com/gradio-app/gradio/pull/3342)
+- Ensure markdown lists are rendered correctly by [@pngwn](https://github.com/pngwn) in [PR 3341](https://github.com/gradio-app/gradio/pull/3341)
```
Looks like there are some additional entries here?

abidlabs commented:
I was taking a look at the kibana and I noticed that the events are being registered, which is awesome! What I don't understand is how to get access to the data about which components are bieng used: 

<img width="853" alt="image" src="https://user-images.githubusercontent.com/1778297/222274763-d698ae58-a269-469e-ac97-5d8065091230.png">

abidlabs commented on gradio/interface.py:
```diff
@@ -324,6 +328,8 @@ def clean_html(raw_html):
             if analytics_enabled is not None
             else os.getenv("GRADIO_ANALYTICS_ENABLED", "True") == "True"
         )
+        if not self.analytics_enabled:
```
We can actually delete all of these lines:

```py
        self.analytics_enabled = (
            analytics_enabled
            if analytics_enabled is not None
            else os.getenv("GRADIO_ANALYTICS_ENABLED", "True") == "True"
        )
        if not self.analytics_enabled:
            os.environ["HF_HUB_DISABLE_TELEMETRY"] = "True"
```

Because `Interface` calls the `Blocks` constructor and passes in `analytics_enabled`

abidlabs commented on gradio/blocks.py:
```diff
@@ -48,6 +48,9 @@
 
 set_documentation_group("blocks")
 
+GRADIO_VERSION = (
```
Just do `from gradio.utils import GRADIO_VERSION` instead of redefining it

abidlabs commented on gradio/interface.py:
```diff
@@ -33,6 +33,10 @@
 
 set_documentation_group("interface")
 
+GRADIO_VERSION = (
```
Just do `from gradio.utils import GRADIO_VERSION` instead of redefining it

abidlabs commented on gradio/utils.py:
```diff
@@ -130,6 +146,65 @@ def launch_analytics_thread(data: Dict[str, Any]) -> None:
 
 
+def launched_telemetry(blocks: gradio.Blocks, data: Dict[str, Any]) -> None:
+    from gradio.blocks import BlockContext
```
Why import inside the function? Better to keep all imports at the top level if possible

dawoodkhan82 commented:
Changed `requirements.txt` to use a specific [commit](https://github.com/huggingface/huggingface_hub/pull/1344/commits/57d23eee58af01611ddade8406e08166db3fcfa1) of `huggingface_hub`. Since I think the plan to release will be in a couple weeks.

abidlabs commented:
> Changed requirements.txt to use a specific [commit](https://github.com/huggingface/huggingface_hub/pull/1344/commits/57d23eee58af01611ddade8406e08166db3fcfa1) of huggingface_hub

It looks like our build process doesn't like this?

dawoodkhan82 commented:
When install the requirements by `pip install -r requirements.txt` it doesn't give that error. But with editable install it's complaining `packaging.requirements.InvalidRequirement: Expected end or semicolon (after name and no valid version specifier)`. Any idea why editable install would be different when installing from a specific git commit @abidlabs?

freddyaboulton commented on test/requirements.in:
```diff
@@ -11,7 +11,7 @@ scikit-image
 pytest
 wandb
-huggingface_hub
+git+https://github.com/huggingface/huggingface_hub.git@57d23eee58af01611ddade8406e08166db3fcfa1
```
This worked for me: 

```
huggingface_hub@git+https://github.com/huggingface/huggingface_hub@57d23eee58af01611ddade8406e08166db3fcfa1
```

Following this guide: https://github.com/pypa/hatch/blob/master/docs/config/dependency.md#version-control-systems

Then you have to add this line to `pyproject.toml`

```
[tool.hatch.metadata]
allow-direct-references = true
```

But shouldn't we just wait for the next hub release?

dawoodkhan82 replied:
I'm in favor of waiting to merge this when `send_telemetry` is released.

freddyaboulton replied:
Word!
